app/models/environment.rb

```diff
@@ -79,7 +79,11 @@ def auto_deploy_enabled?
 
   # The default git ref to deploy when none is provided for this environment.
   def default_ref
-    super.presence || self.class.default_ref
+    client_default_branch || super.presence || self.class.default_ref
+  end
+
+  def client_default_branch
+    @client ||= Octokit.client.repository(repository.name).default_branch
```

Comment:
This is the main change now in the logic. I memoized this in a method call in case the environment calls the `#default_ref` method more than once, so we will only make the request once.